CST-12044 visualize the primary bitstream & CST-12043 primary bitstream flag

vNovski opened the pull request:
**_Please note that this development has been funded by the University of California - California Digital Library._**

## References
DSpace PR: https://github.com/DSpace/DSpace/pull/9195

## Description
- Added a bitstream switch on the download page for each file and also add a switch in the modal editing window of a specific file, which can be used to make the bitstream the primary.
- Added a badge primary aside the file name for the bitstream that has been eventually set as primary

INTERNAL DOCUMENTATION:
- Added a new method to bitstream-data.service.ts findPrimaryBitstreamByItemAndName() which allows you to get a primary bitstream based on item and origin
- In addition to implementing the business logic described in the task, the behavior of the custom switch was also fixed so that it was correctly processed by the form builder, specifically data normalization. Also, the custom switch did not use translate pipe for automatic label translation

## Instructions for Reviewers
Review for primary bitstream switch has been done  with the following cases:

- Edit Submission form

  - Upload files:

    - Scroll down to file list

      - activate the switch for one of the files - **automaticaly saves and updates**

      - activate the switch for another file → **previous primary bitstream  turns off automatically  and new primary bitstream turns on then automaticaly saves and updates**

      - deactivate the switch of current primary bitstream → **primary bitstream  turns off  then automaticaly saves and updates**

      - activate the switch for one of the files - **automaticaly saves and updates**

      - delete file witch is primary -  **file is deleted and primary bitstream is reseted  then automaticaly saves and updates**

    - open file edit modal for one of the files

      - activate the primary bitstream switcher  then click save - **saves and updates**

    - open file edit modal for file witch is not primary

      - activate the primary bitstream switcher then click save - **previous primary bitstream  turns off automatically  and new primary bitstream turns on then  saves and updates**

    - open file edit modal for file witch is primary

      - deactivate the switch of  primary bitstream then click save → **primary bitstream  turns off  then saves and updates**

SCREENSHOTS:
Modal:
![image](https://github.com/DSpace/dspace-angular/assets/33873679/b67ef446-a81f-437a-af97-a1018705f1ee)
Upload Files form: 
![image](https://github.com/DSpace/dspace-angular/assets/33873679/ea14b92f-a758-4d4a-95e0-0df1e6901161)


Review for badge which indicates primary bitstream  has been done  with the following cases:
- if  bitstream is primary  then  **badge [Primary] is displayed** 
![image](https://github.com/DSpace/dspace-angular/assets/33873679/b8a8e619-29b8-40ad-980f-987445fa8def)
- if  bitstream is not primary  then  **badge [Primary] is not displayed** 
![image](https://github.com/DSpace/dspace-angular/assets/33873679/fe34aacc-2c72-4083-a795-b92fcb9761d0)


## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

hardyoyo commented:
I'm attempting to test this PR in concert with the backend PR... and... I cannot see the primary bitstream switch on the submission form. Is there any chance you can post a screenshot of what this is supposed to look like for the submission form (not the edit a bitstream modal, I'm aware of what that modal looks like, and that functionality still works in my testing).

tdonohue commented on src/app/submission/sections/upload/file/section-upload-file.component.html:
```diff
@@ -1,8 +1,21 @@
+    <div class="col-md-10">
       <div class="float-left w-75">
-        <h3>{{fileName}} <span class="text-muted">({{fileData?.sizeBytes | dsFileSize}})</span></h3>
+        <h3>{{fileName}}<span class="text-muted">({{fileData?.sizeBytes | dsFileSize}})</span></h3>
```
It looks like you accidentally removed the space between the filename and the file size.  I don't think we should do that.

tdonohue commented on src/app/submission/sections/upload/section-upload.component.html:
```diff
@@ -26,16 +18,26 @@ <h3 class="text-center"><span class="text-muted">{{'submission.sections.upload.n
-  <ng-container *ngFor="let fileEntry of fileList">
+  <div class="row">
+    <div class="col-md-2">
+      <span class="text-left font-weight-bold">{{ 'bitstream.edit.form.primaryBitstream.label' | translate }}</span>
```
I find the layout of the column to be odd.  It's difficult to tell this is a table (see screenshot).  Should we make sure the files also have a header?  Perhaps make it like this?
```
----------------------------------------------
 Primary File   |    File Details
----------------------------------------------
```

NOTE: I specifically changed the heading to "primary file" as this is called the "Upload files" section.  I'm not sure if users will all understand "primary bitstream"? 

If you feel it should be called "primary bitstream" on the submission form, we can do so.  I'm just not sure if all users will understand that bitstreams and files mean the same thing.

tdonohue reviewed (changes requested):
@vNovski : Thanks!  Tested and reviewed this today.  It looks good overall & works!  That said, I found a small bug, and I think we may want to look at the layout of the page a bit.

Here's what I see on the submission form with this PR (and the backend PR): 
![Primary-bitstream](https://github.com/DSpace/dspace-angular/assets/483997/1ec77cb1-8198-46d1-b53c-a67af280d49b)

A few points about that layout:
* The "Primary Bitstream" header is aligned right, while the buttons are aligned center. It makes it look a bit odd.
* I think we need a header over the file section that says something like "File Details" or "File Information".  That would make it clear this is a table of information.
* (Not Required) I'm starting to wonder if we should rename the "Primary Bitstream" header on the submission form to say "Primary File" or "Primary?".  I noticed that submission form doesn't use the term "bitstsream".  So, we may want to consider just calling everything a "file" on this page.

One small bug that I noticed:  If you click on a primary bitstream, then accept/reject the license, then change the primary bitstream the changes are **not saved**.  Here's how to reproduce:
* Upload at least to bitstreams on the submission form
* Select one as primary (should succeed)
* Now either check or uncheck the license checkbox
* Immediately select the *other* bitstream as the primary bitstream.
* No error will occur, but the primary bitstream is unchanged.  It appears the PATCH request sent was **only** to update the license checkbox and NOT for the new primary bitstream.

I have a few other minor comments inline.  Overall though this all looks good.

vNovski commented:
> @vNovski : Thanks! Tested and reviewed this today. It looks good overall & works! That said, I found a small bug, and I think we may want to look at the layout of the page a bit.
> 
> Here's what I see on the submission form with this PR (and the backend PR): ![Primary-bitstream](https://private-user-images.githubusercontent.com/483997/300971787-1ec77cb1-8198-46d1-b53c-a67af280d49b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyMzAyMzQsIm5iZiI6MTcwNzIyOTkzNCwicGF0aCI6Ii80ODM5OTcvMzAwOTcxNzg3LTFlYzc3Y2IxLTgxOTgtNDZkMS1iNTNjLWE2N2FmMjgwZDQ5Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwNlQxNDMyMTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNTIwOWQzMmVlZjZlZGU4YmRkNTNiOWQ0ZjNlZjFkZTkwMDQxMzcwZmQyYTI1NzQ4ZjE2ODhiYmU5M2NkN2FjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.r4ApoQb_L5I9u8Ukqvmr5IG6GksHwZpYac97i1gJEKg)
> 
> A few points about that layout:
> 
> * The "Primary Bitstream" header is aligned right, while the buttons are aligned center. It makes it look a bit odd.
> * I think we need a header over the file section that says something like "File Details" or "File Information".  That would make it clear this is a table of information.
> * (Not Required) I'm starting to wonder if we should rename the "Primary Bitstream" header on the submission form to say "Primary File" or "Primary?".  I noticed that submission form doesn't use the term "bitstsream".  So, we may want to consider just calling everything a "file" on this page.
> 
> One small bug that I noticed: If you click on a primary bitstream, then accept/reject the license, then change the primary bitstream the changes are **not saved**. Here's how to reproduce:
> 
> * Upload at least to bitstreams on the submission form
> * Select one as primary (should succeed)
> * Now either check or uncheck the license checkbox
> * Immediately select the _other_ bitstream as the primary bitstream.
> * No error will occur, but the primary bitstream is unchanged.  It appears the PATCH request sent was **only** to update the license checkbox and NOT for the new primary bitstream.
> 
> I have a few other minor comments inline. Overall though this all looks good.



For the pull request, I made the following changes:
1. Removed --code-coverage from package.json.
2. Added two texts for the <label class="custom-control-label" for="primaryBitstream{{fileIndex}}"></label> tag, depending on the selected option: "Make {{fileName}} the primary bitstream" and "Remove {{fileName}} as the primary bitstream".
3. Fixed linting issues.
4. Changed the label "Primary bitstream" to "Primary File".
5. Fixed an issue with the license button that was preventing saving, now everything works correctly.

tdonohue commented on src/app/submission/sections/upload/file/section-upload-file.component.html:
```diff
@@ -1,6 +1,22 @@
+        (change)="togglePrimaryBitstream($event)">
+        <label class="custom-control-label" for="primaryBitstream{{fileIndex}}">
+          <span class="sr-only" aria-hidden="true" *ngIf="!isPrimary">{{'submission.sections.upload.primary.make' | translate:{ fileName: fileName } }}</span>
+          <span class="sr-only" aria-hidden="true" *ngIf="isPrimary">{{'submission.sections.upload.primary.remove' | translate:{ fileName: fileName } }}</span>
```
Same here, please remove `aria-hidden=true` attribute

tdonohue reviewed (commented):
Thanks @vNovski !   This looks good except for a **very minor issue**.  I have retested & verified that all prior bugs are fixed and everything else looks good.  

So, once this issue is fixed (and the backend PR is ready), then this will be ready to merge.

vNovski commented:
> > @vNovski : Thanks! Tested and reviewed this today. It looks good overall & works! That said, I found a small bug, and I think we may want to look at the layout of the page a bit.
> > Here's what I see on the submission form with this PR (and the backend PR): ![Primary-bitstream](https://private-user-images.githubusercontent.com/483997/300971787-1ec77cb1-8198-46d1-b53c-a67af280d49b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyMzAyMzQsIm5iZiI6MTcwNzIyOTkzNCwicGF0aCI6Ii80ODM5OTcvMzAwOTcxNzg3LTFlYzc3Y2IxLTgxOTgtNDZkMS1iNTNjLWE2N2FmMjgwZDQ5Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjA2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwNlQxNDMyMTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNTIwOWQzMmVlZjZlZGU4YmRkNTNiOWQ0ZjNlZjFkZTkwMDQxMzcwZmQyYTI1NzQ4ZjE2ODhiYmU5M2NkN2FjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.r4ApoQb_L5I9u8Ukqvmr5IG6GksHwZpYac97i1gJEKg)
> > A few points about that layout:
> > 
> > * The "Primary Bitstream" header is aligned right, while the buttons are aligned center. It makes it look a bit odd.
> > * I think we need a header over the file section that says something like "File Details" or "File Information".  That would make it clear this is a table of information.
> > * (Not Required) I'm starting to wonder if we should rename the "Primary Bitstream" header on the submission form to say "Primary File" or "Primary?".  I noticed that submission form doesn't use the term "bitstsream".  So, we may want to consider just calling everything a "file" on this page.
> > 
> > One small bug that I noticed: If you click on a primary bitstream, then accept/reject the license, then change the primary bitstream the changes are **not saved**. Here's how to reproduce:
> > 
> > * Upload at least to bitstreams on the submission form
> > * Select one as primary (should succeed)
> > * Now either check or uncheck the license checkbox
> > * Immediately select the _other_ bitstream as the primary bitstream.
> > * No error will occur, but the primary bitstream is unchanged.  It appears the PATCH request sent was **only** to update the license checkbox and NOT for the new primary bitstream.
> > 
> > I have a few other minor comments inline. Overall though this all looks good.
> 
> For the pull request, I made the following changes:
> 
> 1. Removed --code-coverage from package.json.
> 2. Added two texts for the  tag, depending on the selected option: "Make {{fileName}} the primary bitstream" and "Remove {{fileName}} as the primary bitstream".
> 3. Fixed linting issues.
> 4. Changed the label "Primary bitstream" to "Primary File".
> 5. Fixed an issue with the license button that was preventing saving, now everything works correctly.

For the pull request, I made the following changes:
1. The aria-hidden attribute has been removed.

tdonohue reviewed (approved):
👍 Thanks @vNovski !

tdonohue commented:
While this has been merged, I'm flagging as `needs documentation` until we can update the Submission documentation to include basic details about this feature: https://wiki.lyrasis.org/display/DSDOC8x/Submission+User+Interface

GraziaQuercia commented:
Hi @tdonohue here's the documentation page: https://wiki.lyrasis.org/display/DSDOC8x/Set+a+bitstream+as+primary
If any change is needed, let me know! Thanks in advance
